modules/hal_rpi_pico/CMakeLists.txt

```diff
@@ -0,0 +1,108 @@
+	elseif(CONFIG_RP2_FLASH_GENERIC_03H)
+		target_sources(boot_stage2 PRIVATE ${BOOT_STAGE_DIR}/boot2_generic_03h.S)
+	elseif(CONFIG_RP2_FLASH_IS25LP080)
+		target_sources(boot_stage2 PRIVATE ${BOOT_STAGE_DIR}/boot2_is25lp080.S)
+	elseif(CONFIG_RP2_FLASH_W25X10CL)
+		target_sources(boot_stage2 PRIVATE ${BOOT_STAGE_DIR}/boot2_w25x10cl.S)
+	else()
+		message(FATAL_ERROR Unknown flash type)
+	endif()
+
+	target_include_directories(boot_stage2 PUBLIC
+		${CMAKE_CURRENT_LIST_DIR}
+		${BOOT_STAGE_DIR}/asminclude
+		${RP2_COMMON_DIR}/pico_platform/include
+		${RP2040_DIR}/hardware_regs/include
+		${COMMON_DIR}/pico_base/include
+		)
+
+	target_link_options(boot_stage2 PRIVATE
+		"-nostartfiles"
+		"--specs=nosys.specs"
+		"LINKER:--script=${BOOT_STAGE_DIR}/boot_stage2.ld"
+		)
+
+	# Generates a binary file from the compiled bootloader
+	add_custom_target(bootloader_bin ALL DEPENDS boot_stage2.bin)
+	add_custom_command(OUTPUT boot_stage2.bin
+		DEPENDS boot_stage2
+		COMMAND ${CMAKE_OBJCOPY} -Obinary $<TARGET_FILE:boot_stage2> boot_stage2.bin
+		)
+
+	# Checksums the binary, pads it, and generates an assembly file
+	add_custom_target(boot_stage2_asm ALL DEPENDS boot_stage2.S)
+	add_custom_command(OUTPUT boot_stage2.S
+		DEPENDS boot_stage2.bin
+		COMMAND ${Python3_EXECUTABLE} ${BOOT_STAGE_DIR}/pad_checksum
+		-s 0xffffffff boot_stage2.bin boot_stage2.S)
+
+	# Pico sources necessary for every build
+	zephyr_library_sources(
+		boot_stage2.S
+		${RP2_COMMON_DIR}/hardware_clocks/clocks.c
+		${RP2_COMMON_DIR}/hardware_pll/pll.c
+		${RP2_COMMON_DIR}/hardware_xosc/xosc.c
+		${RP2_COMMON_DIR}/hardware_watchdog/watchdog.c
+		${RP2_COMMON_DIR}/pico_platform/platform.c
+	)
+
+	# Pico headers necessary for every build
+	zephyr_include_directories(
+		${RP2_COMMON_DIR}/hardware_base/include
+		${RP2_COMMON_DIR}/hardware_clocks/include
+		${RP2_COMMON_DIR}/hardware_gpio/include
+		${RP2_COMMON_DIR}/hardware_watchdog/include
+		${RP2_COMMON_DIR}/hardware_xosc/include
+		${RP2_COMMON_DIR}/hardware_pll/include
+		${RP2_COMMON_DIR}/hardware_vreg/include
+		${RP2_COMMON_DIR}/hardware_irq/include
+		${RP2_COMMON_DIR}/hardware_sync/include
+		${RP2_COMMON_DIR}/hardware_timer/include
+		${RP2_COMMON_DIR}/hardware_resets/include
+		${RP2040_DIR}/hardware_regs/include
+		${RP2040_DIR}/hardware_structs/include
+		${COMMON_DIR}/pico_base/include
+		${COMMON_DIR}/pico_binary_info/include
+		${RP2_COMMON_DIR}/pico_platform/include
+		${CMAKE_CURRENT_LIST_DIR}
+	)
+
+	macro(rpi_pico_simple_hardware_target name)
+		zephyr_library_sources(${RP2_COMMON_DIR}/hardware_${name}/${name}.c)
+		zephyr_include_directories(${RP2_COMMON_DIR}/hardware_${name}/include)
+	endmacro()
+
+	if(CONFIG_GPIO OR CONFIG_PINCTRL)
+		rpi_pico_simple_hardware_target(gpio)
+	endif()
+
+	if(CONFIG_SERIAL)
+		rpi_pico_simple_hardware_target(uart)
+	endif()
+
+	if(CONFIG_ADC)
+		rpi_pico_simple_hardware_target(adc)
+	endif()
+
+	if(CONFIG_PWM)
+		rpi_pico_simple_hardware_target(pwm)
+	endif()
```

Comment:
This needs to be improved:

- Create Kconfig options for each HAL module, e.g.

```
config PICOSDK_USE_GPIO
    bool
    help
        ...
```
- You may default to `y` those that are always needed
- Use this information to compile the HAL efficiently:

```
zephyr_library_sources_ifdef(CONFIG_PICOSDK_USE_GPIO ${RP2_COMMON_DIR}/hardware_gpio/gpio.c)
zephyr_include_directories_ifdef(CONFIG_PICOSDK_USE_GPIO ${RP2_COMMON_DIR}/hardware_gpio/include)
```
- Drivers need to select required HAL components:

```
config PINCTRL_RPI_PICO
    bool "Pinctrl driver for Raspberry Pi Pico"
    select PICOSDK_USE_GPIO
    help
      ...
```